Fix pip.installed state managed pre-release upgrades

kurtrwall opened the pull request:
### What does this PR do?

### What issues does this PR fix or reference?
Fixes #68525 

### Previous Behavior
Python packages installed in an environment would not upgrade to their pre-release version when any exist in the index.

### New Behavior
Pre-release versions of packages managed by `pip.installed` state are upgraded when using `upgrade=True` and `pre_releases=True`.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

<!-- Please review Salt's Contributing Guide for best practices and guidance in
choosing the right branch:
https://docs.saltproject.io/en/master/topics/development/contributing.html -->

<!-- Additional guidance for pull requests can be found here:
https://docs.saltproject.io/en/master/topics/development/pull_requests.html -->

<!-- See GitHub's page on GPG signing for more information about signing commits
with GPG:
https://help.github.com/articles/signing-commits-using-gpg/ -->

kurtrwall committed:
Fix pip.installed state managed pre-release upgrades

`pip index versions` will never yield pre-release versions of the
package without the `--pre` option.

In effect, this means providing `pre_releases=True` and `upgrade=True`
to the `pip.installed` state will never work, but I'm unsure of the
actual extent of this bug.

kurtrwall commented:
I've been able to trace the bug back to the inception of the code that uses [pip index versions](https://github.com/saltstack/salt/commit/eaf061a688646518a124965f797fa1a7f99431cb), so I based the PR on the v3004.x branch, but I'm unsure if that was the appropriate one.

I'm also having a bit of trouble with the tests, but I didn't want it to block the contribution.

kurtrwall committed:
Add changelog file

kurtrwall committed:
Fix test logic for failing platforms

kurtrwall committed:
Fix test

twangboy commented:
Just install pre-commit (`pip install pre-commit`)
Add your changed files (`git add .`)
And run pre-commit (`pre-commit`)

kurtrwall commented:
Okay, sounds like I had the right setup for the most recent push. Glad to see it's approved! Thank you.

twangboy commented:
still need to see passing tests

kurtrwall committed:
Fix test argument

kurtrwall committed:
Add mocking for the prerelease tests

This is bit ugly, but seemingly some platforms arent receiving the same versions from PyPI for typing-extensions.

kurtrwall commented:
It seems like the only failing test was due to a timeout, is there a way to rerun just the one test?

kurtrwall commented:
Thanks for the approval @twangboy! I'm sure we'd like to get this merged and on the way to release as quickly as possible. I do notice there are still some checks failing, so I just wanted to offer I'm ready to help get things resolved. That said, I'll be unavailable starting next week until June. Thanks again!

Commit:
Regenerate static lockfiles after adding packaging to pytest.txt
